### PR TITLE
HTMLLintBear: Add `use_spaces` option

### PR DESCRIPTION
**Pull request by aashraybhandar1**
Fixes issue #1741

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: bears/hypertext/HTMLLintBear.py [hunk 1]
```diff
@@ -27,10 +27,13 @@ class HTMLLintBear:
...
     @staticmethod
-    def create_arguments(filename, file, config_file,
+    def create_arguments(filename, file, config_file, use_spaces: bool,
                          htmllint_ignore: typed_list(str)=()):
         """
         :param htmllint_ignore: List of checkers to ignore.
+        :param use_spaces: True if spaces are to be used instead of tabs.
         """
         ignore = ','.join(part.strip() for part in htmllint_ignore)
+        if(not use_spaces):
```
Comment (Makman2) [inline bears/hypertext/HTMLLintBear.py]: unnecessary braces, this is not C ;)
```
if not use_spaces:
```

**Comment by Makman2** [comment]
Please look at our commit guidelines on how to phrase messages.

So:
1. Use imperative tense: Implemented -> Implement
2. Actually I would remove the commit body (except the issue reference), I don't understand what you try to say with it.

Also you have to write a test testing your new setting. Also the current tests fail, this needs to be fixed :+1:
(Probably because `use_spaces` became a mandatory setting, you have to explicitly define it in every test now)

**Comment by RaiVaibhav** [comment]
@aashraybhandar1 add the issue link also in comment not only in commit message

File: bears/hypertext/HTMLLintBear.py [hunk 2]
```diff
@@ -27,10 +27,13 @@ class HTMLLintBear:
...
-    def create_arguments(filename, file, config_file,
+    def create_arguments(filename, file, config_file, use_spaces: bool,
                          htmllint_ignore: typed_list(str)=()):
         """
         :param htmllint_ignore: List of checkers to ignore.
+        :param use_spaces: True if spaces are to be used instead of tabs.
         """
         ignore = ','.join(part.strip() for part in htmllint_ignore)
+        if not use_spaces:
+            ignore += ',tabs'
```
Comment (Makman2) [inline bears/hypertext/HTMLLintBear.py]: what if `htmllint_ignore` is empty? then your string has a leading comma:
```
>>> ignore
',tabs'
```
I think cleanest is to embed this into the `join`, maybe like this:
```python
additional_ignore = []
if not use_spaces:
    additional_ignore.append('tabs')
ignore = ','.join(part.strip() for part in htmllint_ignore + additional_ignore)
```

File: tests/hypertext/HTMLLintBearTest.py [hunk 3]
```diff
@@ -9,21 +9,47 @@
...
+    <p class='foo'>
+    <h1>Hello, world!</h1>
+    <br> </br>
+  </body>
+</html>
+"""
+
 HTMLLintBearTest = verify_local_bear(HTMLLintBear,
-                                     valid_files=(),
+                                     valid_files=('',),
```
Comment (Makman2) [inline tests/hypertext/HTMLLintBearTest.py]: Is this necessary for the feature? Otherwise different commit, testing an empty string is definitely not bad :+1:

File: tests/hypertext/HTMLLintBearTest.py [hunk 4]
```diff
@@ -9,21 +9,47 @@
 </html>
 """
 
+invalid_test = """
```
Comment (Makman2) [inline tests/hypertext/HTMLLintBearTest.py]: could you maybe name it a bit more precise? Usually it's always: Give a name that resembles the usage. Like `file_with_invalid_spacing` or so

File: bears/hypertext/HTMLLintBear.py (hunk 2 above)
Comment (aashraybhandar1) [reply bears/hypertext/HTMLLintBear.py]: I will make the necessary changes :+1:

File: bears/hypertext/HTMLLintBear.py [hunk 5]
```diff
@@ -27,10 +27,15 @@ class HTMLLintBear:
...
-    def create_arguments(filename, file, config_file,
-                         htmllint_ignore: typed_list(str)=()):
+    def create_arguments(filename, file, config_file, use_spaces: bool,
+                         htmllint_ignore: typed_list(str)=[]):
         """
         :param htmllint_ignore: List of checkers to ignore.
+        :param use_spaces: True if spaces are to be used instead of tabs.
         """
+        if not use_spaces:
+            htmllint_ignore.append('tabs')
```
Comment (Makman2) [inline bears/hypertext/HTMLLintBear.py]: this is what we shouldn't do: You potentially modify the default value for `htmllint_ignore`. Consecutive calls to `create_arguments` might put the item `tabs` multiple times into the list. You have to create a new list for that^^

File: bears/hypertext/HTMLLintBear.py [hunk 6]
```diff
@@ -27,10 +27,15 @@ class HTMLLintBear:
...
+
         ignore = ','.join(part.strip() for part in htmllint_ignore)
+
```
Comment (Makman2) [inline bears/hypertext/HTMLLintBear.py]: unnecessary change

**Comment by aashraybhandar1** [comment]
@Makman2 are the changes fine?

File: tests/hypertext/HTMLLintBearTest.py (hunk 3 above)
Comment (Makman2) [reply tests/hypertext/HTMLLintBearTest.py]: bump :)

**Comment by Makman2** [comment]
See https://github.com/coala/coala-bears/pull/2073/commits/ee4b91d7181a97bb0f401be13d6b19e7ddaaf5b9#r145260641 ;)

File: tests/hypertext/HTMLLintBearTest.py (hunk 3 above)
Comment (aashraybhandar1) [reply tests/hypertext/HTMLLintBearTest.py]: I'm sorry I didn't get the meaning of the last comment.. should I remove the changes mentioned above regarding the empty string, make another issue and implement the same there?

Comment (Makman2) [reply tests/hypertext/HTMLLintBearTest.py]: just curious why you now introduced this empty string. If this can be separated into a new commit, this would ease review.

Comment (aashraybhandar1) [reply tests/hypertext/HTMLLintBearTest.py]: In https://github.com/coala/coala-bears/pull/1304/files one of the reviewers mentioned it's bad practice to leave the parameter for valid/invalid file to be empty.Hence I changed that here, no other particular reason

Comment (Makman2) [reply tests/hypertext/HTMLLintBearTest.py]: Then, move this into a new commit please :D You can just put it ontop in this PR :+1:

Comment (Makman2) [reply tests/hypertext/HTMLLintBearTest.py]: Exactly write this reason into the new commit message ;D

**Comment by aashraybhandar1** [comment]
@Makman2  are the changes fine ?

File: tests/hypertext/HTMLLintBearTest.py [hunk 7]
```diff
@@ -9,21 +9,45 @@
...
+test_ignore_tabs = """
+<html>
+	<body>
+		<h1>Hello, world!</h1>
+	</body>
+</html>
+"""
+
+HTMLLintBearIgnoreTabs = verify_local_bear(
+    HTMLLintBear,
```
[the remark this answers is not in the document]
Comment (stellargo) [reply tests/hypertext/HTMLLintBearTest.py]: @aashraybhandar1 gotta fix this :)

File: tests/hypertext/HTMLLintBearTest.py [hunk 8]
```diff
@@ -12,18 +12,34 @@
...
 
 HTMLLintBearIgnoreQuotationTest = verify_local_bear(
     HTMLLintBear,
     valid_files=(),
     invalid_files=(test_file,),
-    settings={'htmllint_ignore': 'quotation'},
+    settings={'use_spaces': 'False', 'htmllint_ignore': 'quotation'},
+    tempfile_kwargs={'suffix': '.html'})
+
+test_ignore_tabs = """
```
Comment (stellargo) [inline tests/hypertext/HTMLLintBearTest.py]: I think it it would be a better idea if you keep your files together (at the top with test_file) :)

File: bears/hypertext/HTMLLintBear.py [hunk 9]
```diff
@@ -27,10 +27,17 @@ class HTMLLintBear:
     CAN_DETECT = {'Syntax', 'Formatting'}
 
     @staticmethod
-    def create_arguments(filename, file, config_file,
-                         htmllint_ignore: typed_list(str)=()):
```
Comment (stellargo) [inline bears/hypertext/HTMLLintBear.py]: I am not sure but I think you should do this in a separate commit, isnt exactly the issue you are resolving here :) You need to add a commit body too :)

Comment (aashraybhandar1) [reply bears/hypertext/HTMLLintBear.py]: Some of the tests aren't passing if this change isn't made.Previously htmllint_ignore acts as a tuple if no argument is passed and is typecasted into a list when arguments are passed which is creating problems in order to append the same.

Comment (stellargo) [reply bears/hypertext/HTMLLintBear.py]: Ok... That fix can be merged before the rest, @Makman2 your thoughts ?

Comment (stellargo) [reply bears/hypertext/HTMLLintBear.py]: But should there be a problem putting the change in a different commit?

Comment (aashraybhandar1) [reply bears/hypertext/HTMLLintBear.py]: In the same issue two different commits or should I create another issue?

Comment (stellargo) [reply bears/hypertext/HTMLLintBear.py]: two commits with each linked to a different issue, but wait for Makman2 to weigh in on this

File: bears/hypertext/HTMLLintBear.py [hunk 10]
```diff
@@ -13,7 +13,6 @@
 class HTMLLintBear:
     """
     Check HTML source code for invalid or misformatted code.
-
```
Comment (Makman2) [inline bears/hypertext/HTMLLintBear.py]: irrelevant change

File: bears/hypertext/HTMLLintBear.py (hunk 9 above)
Comment (Makman2) [reply bears/hypertext/HTMLLintBear.py]: imo that's fine, it's a consequence to use a list instead of a tuple to implement the desired features. In fact it's better to use a tuple to make the default value immutable, however this isn't really practical here, we have to use a list.

**Comment by Makman2** [comment]
nearly ready, just remove that unnecessary line-deletion ;)

**Comment by aashraybhandar1** [comment]
Are the changes fine now :grimacing:

**Comment by Makman2** [comment]
Could you just add an empty line in your commit body between your description and the `Closes xxxx` reference? :)

**Comment by aashraybhandar1** [comment]
@Makman2  I can't understand why these two tests failed.Could you guide me with what went wrong?

**Comment by Makman2** [comment]
> codecov/project — 99.68% (target 100%) 

You have not enough tests. Please see the build logs which lines are not covered.

**Comment by RaiVaibhav** [comment]
@aashraybhandar1 @Makman2  `Travis CI` fail was  bug :sweat_smile: 
@aashraybhandar1 Repush without any change, It will pass.

**Comment by aashraybhandar1** [comment]
@Makman2  as far as I could understand it's showing that with respect to tests in the JS folder for the ESLintBear. It is showing 100%for the HyperText folder

**Comment by Makman2** [comment]
ack cf2f3f4

**Comment by aashraybhandar1** [comment]
Thank you @Makman2  :smiley: .So now I have to just wait for it to get merged?

**Comment by aashraybhandar1** [comment]
Can someone merge this pull request :disappointed:

**Comment by Makman2** [comment]
@gitmate-bot rebase

**Comment by Makman2** [comment]
@gitmate-bot rebase

**Comment by aashraybhandar1** [comment]
@Makman2  I did the manual rebase

**Comment by Makman2** [comment]
@rultor merge

**Comment by Makman2** [comment]
@gitmate-bot rebase

**Comment by Makman2** [comment]
Let's do a manual merge now ;)

**Comment by aashraybhandar1** [comment]
@Makman2  thanks a lot :D
